### DISCUSSION TITLE
How can bundles programmatically define DataObject Classes?

### DISCUSSION THREAD

**Discussion by mlkdev** [pimcore/pimcore#10414] [Q&A / Support]
I'm working on a custom bundle which is primarily a package of templates with areas and areabricks. In some cases, these areabricks require the use of DataObjects, which outside of bundle development you would create through the Pimcore admin UI.

If possible, my question is how do bundles define DataObject Classes that are compatible with the right-click > Add Object menu in the Pimcore DataObject UI?

My understanding is that Pimcore is designed to look specifically for `var/classes/definition_*.php` files to construct the classes which are ultimately what developers will interface with when writing controllers and/or services. I've attempted to move these definitions and class constructor files over into the bundle after creating them through the Pimcore UI, which did not work.

Any direction on how to achieve this or should I stick to the Pimcore UI for now?

---

**Update:**

Continuing my experimentation. It looks like I can edit the `/config/pimcore/constants.php` file to define the class folder, along with changing the `"\\Pimcore\\Model\\DataObject\\"` value in the composer file. This allows me to build and manage the classes within my bundle as desired, though I want to confirm that this is the best approach.

**Comment by mjnewman** [comment]
@mlkdev On the previous Pimcore discussion board it was [recommended to add data object classes during the bundle installation](https://talk.pimcore.org/t/custom-bundle-location-of-class-definitions/2517).

An example was shown using the [Ecommerce Bundle](https://github.com/pimcore/pimcore/blob/10.x/bundles/EcommerceFrameworkBundle/Tools/Installer.php#L210) – You can see the JSON files imported by the installer [here](https://github.com/pimcore/pimcore/tree/10.x/bundles/EcommerceFrameworkBundle/Resources/install/class_sources).

During development, you could create your classes using the CMS and then export the JSON data from the CMS at **Settings > Data Objects > Bulk Export** 

Once exported, it's probably necessary to remove some parameters such as id, userOwner, etc.

The JSON could be added to your bundle and the classes would be created during bundle installation similar to the Ecommmerce Bundle example.

Can anyone confirm if this is the best approach or advise on other ways to achieve this? We're looking to do something similar and want to confirm.
